src/devices/kedsum.c

```diff
@@ -41,44 +51,33 @@ static int kedsum_callback(r_device *decoder, bitbuffer_t *bitbuffer) {
     // require at least 4 received repeats
     int r = bitbuffer_find_repeated_row(bitbuffer, 4, 42);
-    if (r<0 || bitbuffer->bits_per_row[r] != 42)
+    if (r < 0 || bitbuffer->bits_per_row[r] != 42)
         return 0;
 
-    uint8_t *b = bb[r];
-
-    uint8_t humidity;
-    uint8_t channel;
-    uint16_t temperature_with_offset;
-    float temperature_f;
+    // remove the two leading 0-bits and align the data
+    bitbuffer_extract_bytes(bitbuffer, r, 2, b, 40);
 
-    channel  = (uint8_t)(((b[1] & 0x0C) >> 2) + 1);
-    humidity = (uint8_t)((b[3] & 0x03) << 6) | ((b[4] & 0xC0) >> 2) | ((b[3] & 0x3C) >> 2);
-
-    uint8_t tnH, tnM, tnL;
-    tnL = ((b[1] & 0x03) << 2) | ((b[2] & 0xC0) >> 6); // Low temp nibble
-    tnM = ((b[2] & 0x3C) >> 2);                        // Med temp nibble
-    tnH = ((b[2] & 0x03) << 2) | ((b[3] & 0xC0) >> 6); // high temp nibble
+    int chk = lfsr4(b, 36, 0x3);
+    if (chk != (b[4] & 0xf))
+        return 0;
 
-    temperature_with_offset =  (tnH<<8) | (tnM<<4) | tnL;
-    temperature_f = (temperature_with_offset - 900) / 10.0;
+    int id       = b[0];
+    int battery  = b[1] >> 6;
+    int channel  = ((b[1] & 0x30) >> 4) + 1;
+    int temp_raw = ((b[2] & 0x0f) << 8) | (b[2] & 0xf0) | (b[1] & 0x0f);
+    int humidity = ((b[3] & 0x0f) << 4) | ((b[3] & 0xf0) >> 4);
+    float temp_f = (temp_raw - 900) * 0.1;
 
-    if (decoder->verbose) {
-      fprintf(stdout, "Bitstream HEX        = ");
-      bitrow_print(b, 48);
-      fprintf(stdout, "Humidity HEX         = %02x\n", b[3]);
-      fprintf(stdout, "Humidity DEC         = %u\n",   humidity);
-      fprintf(stdout, "Channel HEX          = %02x\n", b[1]);
-      fprintf(stdout, "Channel              = %u\n",   channel);
-      fprintf(stdout, "temp_with_offset HEX = %02x\n", temperature_with_offset);
-      fprintf(stdout, "temp_with_offset     = %d\n",   temperature_with_offset);
-      fprintf(stdout, "TemperatureF         = %.1f\n", temperature_f);
-    }
+    char *battery_str = battery == 2 ? "OK" : battery == 1 ? "WEAK" : "LOW";
 
     data = data_make(
-            "model",         "",            DATA_STRING, "Kedsum Temperature & Humidity Sensor",
-            "channel",       "Channel",     DATA_INT, channel,
-            "temperature_F", "Temperature", DATA_FORMAT, "%.02f F", DATA_DOUBLE, temperature_f,
-            "humidity",      "Humidity",    DATA_FORMAT, "%u %%", DATA_INT, humidity,
+            "model",            "",             DATA_STRING, "Kedsum-TH",
+            "id",               "ID",           DATA_INT, id,
+            "channel",          "Channel",      DATA_INT, channel,
+            "battery",          "Battery",      DATA_STRING, battery_str,
+            "temperature_F",    "Temperature",  DATA_FORMAT, "%.02f F", DATA_DOUBLE, temp_f,
+            "humidity",         "Humidity",     DATA_FORMAT, "%u %%", DATA_INT, humidity,
+            "mic",              "Integrity",    DATA_STRING, "CRC",
```

Comment:
s/CRC/LFSR/ ? It is technically not a crc per my understanding.

Reply:
Not technically a proper "CRC". But it's of comparable strength. I'd say the difference is that a CRC is strictly linear an a general LFSR is not. This one in particular looks linear though :) It seems a "degenerate" (fixed key) form of the common digest type -- or maybe I'm just not seeing how this is perhaps a broken CRC-4. I definitly going to investigate the algorithm further.
Currently the MIC has values of `PARITY < CHECKSUM < CRC` relating to strength. Perhaps we could suffix a width, e.g. `PARITY-1, PARITY-2, SUM-8 < CRC-4, CRC-8, CRC-16` types should cover it.

Reply:
I leave it up to you to decide. It is also fine as it is. The current model is sort of broken as some CRC's are 4 bits and others are 16 bits. The difference is significant with regards to that.

Reply:
Ah well, turns out it really is a CRC-4. But the last 4 bits are only folded in, not processed. I can imagine the faulty CRC lagging 4 bits in the processing. Could also be by design but that would be a poor choice.